### DISCUSSION TITLE
Clarifications regarding using JBANG with Tomcat 9 and JAR files

### DISCUSSION THREAD

**Discussion by nimayneb** [jbangdev/jbang#1858] [Q&A] [answered]
Dear JBANG Team,

I have a few questions and would appreciate your guidance on the following:

1. Tomcat 9 and Application (compiled into single-JAR?): Are there any known best practices or experiences with using JBANG for a Tomcat 9 application that is compiled into a JAR file? Are there any particular considerations I should be aware of?
2. JAR File Structure: I’ve noticed that when creating a JAR file with JBANG, the "classes" directory is missing, and all classes files with packages are placed directly in the root of the JAR. Is this structure compatible with Tomcat, or could it potentially cause issues?
3. Including all Dependencies in the JAR: How can I include all required dependencies in the JAR file? Is this only possible via the Export-FATJAR function, or can it also be done directly during the "Run" or "Build" process?

Here’s the Java file (without Java) I am currently using:
```
///usr/bin/env jbang "$0" "$@" ; exit $?

//JAVA 17

/**
 * Tomcat Dependencies
 */

//DEPS org.apache.tomcat:tomcat-catalina:9.0.96
// and so on...

/**
 * Dependencies
 */
 
//DEPS ...
// and so on...

// Files
// -------
//RUNTIME_OPTIONS -Djava.util.logging.config.file=/dev/project/tomcat/conf/logging.properties
//RUNTIME_OPTIONS -Dlog4j.configurationFile=/dev/project/tomcat/webapps/project/WEB-INF/log4j2.xml

// Classes
// -------
//RUNTIME_OPTIONS -Djava.util.logging.manager=org.apache.juli.ClassLoaderLogManager
//RUNTIME_OPTIONS -Djava.protocol.handler.pkgs=org.apache.catalina.webresources

// Paths
// -------
//RUNTIME_OPTIONS -Dcatalina.base=/dev/project/tomcat
//RUNTIME_OPTIONS -Dcatalina.home=/dev/project/tomcat
//RUNTIME_OPTIONS -Djava.io.tmpdir=/dev/project/tomcat/temp
//RUNTIME_OPTIONS -Dignore.endorsed.dirs=

// Values
// -------
//RUNTIME_OPTIONS -Djdk.tls.ephemeralDHKeySize=2048

//SOURCES ../src/**/*.java

//MAIN org.apache.catalina.startup.Bootstrap
```
I look forward to your feedback and appreciate your support in advance.

Best regards,
Jan

**Comment by quintesse** [comment]
Hi @nimayneb , I don't think anyone of us has ever tried using an embedded Tomcat with JBang, so this is unknown territory for us as well.

Regarding the second point I can tell you that JARs normally don't have a `classes` folder, so if that's required for Tomcat it's most likely because it uses some kind of custom class loader. JBang doesn't support generating any kind of custom folder structure in JARs so if that's required you couldn't use JBang to create the JAR. But perhaps Tomcat will work using a normal JAR file, but that depends on Tomcat, you'd have to try and see if it works or not.

And finally, yes, `export fatjar` is the only way to create a JAR with dependencies included. But I have to warn you that the feature is experimental and its implementation is pretty simplistic so I don't know if it will work for something relatively complex as an embedded web server.

PS: IMO JBang is really good at _running_ stuff, you can point it at files, JARs, Maven artifacts, both locally and remote and it will try to run them. But JBang is not really meant for _building_ projects that result in distributable binaries. That's where Maven and Gradle shine.

**Comment by quintesse** [comment]
Btw, we do have several examples of other embedded web servers here: https://github.com/jbangdev/jbang-examples/tree/main/examples

You could take a look at `javalin.java`, `jetty.jsh`, `ratpack.java`, `sparkjava.java`, `undertow.java` and even `vertx.java`.
But those are of course all simple web servers, not app servers, so that might not be what you're looking for.

**Comment by quintesse** [comment]
And as a final comment, if there is a way to run your Tomcat application in any way like the examples I gave in the previous example: by writing code that sets up a Tomcat instance and if you can tie that directly to your application code then I think the folder structure of the JAR shouldn't matter and you should be able to use the JAR that JBang generates. The only remaining issue would be to see if the `export fatjar` would work.

**Comment by maxandersen** [comment, marked as answer]
I was curious - so as @quintesse says then making custom archives is out of scope for jbang but embedded usecase should work.

Here is what I got working with some StackOverflow and chatgpt help :)

```
///usr/bin/env jbang "$0" "$@" ; exit $?
//DEPS org.apache.tomcat.embed:tomcat-embed-core:10.1.0
//DEPS org.apache.tomcat.embed:tomcat-embed-jasper:10.1.0
//DEPS javax.servlet:javax.servlet-api:4.0.1

import java.io.File;
import java.io.IOException;
import java.io.PrintWriter;

import org.apache.catalina.Context;
import org.apache.catalina.LifecycleException;
import org.apache.catalina.startup.Tomcat;

import jakarta.servlet.ServletException;
import jakarta.servlet.http.HttpServlet;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;

public class App {

    public static void main(String[] args) throws LifecycleException {
        Tomcat tomcat = new Tomcat();
        tomcat.setBaseDir("temp");
        tomcat.setPort(8089);
        tomcat.getConnector();

        String contextPath = "";
        String docBase = new File(".").getAbsolutePath();

        Context context = tomcat.addContext(contextPath, docBase);

        class SampleServlet extends HttpServlet {

            @Override
            protected void doGet(HttpServletRequest req, HttpServletResponse resp)
                    throws ServletException, IOException {
                PrintWriter writer = resp.getWriter();

                writer.println("<html><title>Welcome</title><body>");
                writer.println("<h1>Have a Great Day!</h1>");
                writer.println("</body></html>");
            }
        }

        String servletName = "SampleServlet";
        String urlPattern = "/aa";

        tomcat.addServlet(contextPath, servletName, new SampleServlet());
        context.addServletMappingDecoded(urlPattern, servletName);

        tomcat.start();
        tomcat.getServer().await();
    }
}
```

then visit http://localhost:8089/aa 

hope that helps.

**Reply by quintesse** [reply]
ChatGPT FTW!

Btw, did you try to create an uberjar and run it that way?

**Reply by nimayneb** [reply]
Thank you for the many replies and the efforts! 😄 Yes, I’ve ignored the fact that JBang doesn’t really position itself as a tool for setting up projects. But honestly, I see the potential for it to handle this anyway—thanks to its simplicity.

I have a strong aversion to the hassle and complexity of Gradle or Maven projects. It feels like they go out of their way to overcomplicate even the simplest tasks. 😅 All I need is something straightforward, like JBang, that lets me spin up a server or whatever without juggling a dozen tech stacks.

And yes, the approach suggested by @maxandersen is the one I'll follow. So use Java to create the exact configurations that Tomkitty tries to abstract into web.xml/server.xml/conf.xml. After all, there are quite a few environments and configurations required before it even works properly. I'll try to keep you updated if my efforts bear some juicy fruits. 🍇

Many thanks to you all for your work on this Java Runner! 🚀
